modules/debugging/index.qmd

```diff
@@ -95,29 +94,271 @@ detach_package(tidyverse)
+
+**Trusty (sometimes) AI.** 
+https://www.infoworld.com/article/3694611/8-chatgpt-tools-for-r-programming.html
+
+https://www.theregister.com/2024/01/27/ai_coding_automatic/
+
+::: callout-tip
+## Help and Learn
+
+Answering questions on the forum can help consolidate your understanding and prepare you for asking and answering questions in other forums such as Stack Overflow.
+:::
+
+**Teachers and Tutors in Consultations.** Ask your lecturers and tutors for help during consultation sessions, which are available for any content-related difficulties, including debugging code.
+
+-   Prepare a small demonstration of your error beforehand to make the session more effective.
+-   The benefit of these sessions is that we can guide you closer to the solution until you figure it out yourself.
+-   Don't be disheartened if your error, which you've struggled with for hours, is solved in minutes — we've had plenty of practice (also being stuck for hours) troubleshooting these errors.
+
+# Minimal Reproducible Examples
+
+::: callout-note
+## Remember
+
+A minimally reproducible example (often called an MRE) contains all of the code necessary for someone else to encounter the same error you're facing.
+:::
+
+To create one, you need to provide the **exact packages you're using**, some **data** if needed, and **your code.** A good example is also **minimal**, which means that you've taken the time to remove as much unrelated packages, code and data as possible to produce the error.
+
+## Providing minimal datasets
+
+If your question involves using a dataset, you will need to provide a small dataset to demonstrate the problem. Often people say that the data is confidential and can’t be provided, but usually the problem isn’t specific to the data you are using. Taking the time to try your code on a different dataset is a good debugging step, and R provides many good datasets to use in these examples built in. You can find a complete list of datasets available to all R users with the `data(package = "datasets")` function.
+
+If you do need to use a custom dataset specific to your project, try reducing the size of the dataset before making your example. The error may only occur for some specific rows of your data. Once you've narrowed in on the issue, you can convert your minimal data object to code for producing it using the `dput()` function.
+
+For example you can use use the below code to provide the first 6 rows of the iris dataset.
+
+```{r}
+dput(head(iris))
+```
+
+Try it yourself! Create a minimal dataset using the `dput()` function for the `penguins` dataset.
+
+::: callout-tip
+## Hint
+
+We want to be putting the `penguins` data to create a minimal dataset.
+:::
+
+```{webr-teachr}
+## REVISIT THIS :( 
+# Load the penguins dataset
+library(palmerpenguins)
+# Use dput() to create a minimal dataset
+<<dput(head(penguins))>>
   
 ???
   
+"Did you use dput() with head(penguins)?" != search_ast(.code, .expr = dput(head(penguins)))
```

Comment:
Two things:

1. These checks need to be in a named vector, i.e. created with `c()`.
2. `!=` won't work here like that, with this you're comparing the hint `"Did you use dput() with head(penguins)?"` with the result of `search_ast(.code, .expr = dput(head(penguins)))` which is TRUE or FALSE. As a result, you always get `FALSE`.

   Instead we have `c("hint" = <lgl>)`.
   
```suggestion
c(
  "Did you use dput() with head(penguins)?" = !search_ast(.code, .expr = dput(head(penguins)))
)
```

If you're stuck with things like this, just tag me in the PR with @mitchelloharawild (help!)